File: budget_app/src/services/user_service.rs
use crate::models::{User, NewUser};
use crate::schema::*;
use crate::schema::users::id;
use diesel::sqlite::SqliteConnection;
use diesel::prelude::*;

use self::users::{pw_hash, username};

pub fn create_user(conn: &mut SqliteConnection, new_user: NewUser) -> Result<usize, diesel::result::Error> {
    diesel::insert_into(users::table)
        .values(new_user)
        .execute(conn)
}

pub fn get_user_by_id(conn: &mut SqliteConnection, user_id: i32) -> Result<User, diesel::result::Error> {
    users::table.find(user_id).first(conn)
}

pub fn delete_user(conn: &mut SqliteConnection, target_user_id: i32) -> Result<usize, diesel::result::Error> {
    diesel::delete(users::table.filter(id.eq(target_user_id))).execute(conn)
}

pub fn get_user_by_credentials(conn: &mut SqliteConnection, input_username: String, input_pw: String) -> Result<User, diesel::result::Error> {
    users::table.find(id)
        .filter(username.eq(input_username))
        .filter(pw_hash.eq(input_pw))
        .first(conn)
}

pub fn get_user_by_username(conn: &mut SqliteConnection, target_username: &str) -> Result<User, diesel::result::Error> {
    users::table.filter(username.eq(target_username)).first(conn)
}

pub fn get_users(conn: &mut SqliteConnection) -> Result<Vec<User>, diesel::result::Error> {
    users::table.load(conn)
}

File: budget-app-api/services/user_service.py
from models.models import db, User
from werkzeug.security import check_password_hash

def create_user(username, pw_hash):
    try:
        new_user = User(username=username, pw_hash=pw_hash)
        db.session.add(new_user)
        db.session.commit()
        return True
    except Exception as e:
        return str(e)

def get_users():
    try:
        return (True, db.session.query(User).all())
    except Exception as e:
        return (False, str(e))

def get_user_by_username(username):
    try:
        return db.session.query(User).filter_by(username=username).first()
    except Exception as e:
        return str(e)
    
def get_user_by_id(user_id):
    try:
        return db.session.query(User).filter_by(id=user_id).first()
    except Exception as e:
        return None
    
def validate_login(username, password):
    try:
        user = get_user_by_username(username=username)
        if not check_password_hash(pwhash=user.pw_hash, password=password):
            raise ValueError("Passwords do not match") 
        return (user, None)
    except Exception as e:
        return (None, str(e))